_develop/keypress-on-entry.md

```diff
@@ -0,0 +1,114 @@
+---
+layout: develop
+title: Respond To Specific Keypress
+
+summary: As a follow up to the writing widgets tutorial, this page will go through how to extend an Entry widget and respond to a specific keypress in it. At the end of the tutorial, you will have created an Entry that responds to pressing the Enter/Return button on the keyboard.
+
+order: 120
+---
+
+## Respond To Specific Keypress In Entry
+---
+
+In the traditional sense, GUI programs have used callbacks to customize actions for widgets. Fyne does not expose inserting closure callbacks inside widgets, but it does not need to. The Go language is plenty extensible to make this work.
+
+Simply use Type Embedding and extend the widget, along with adding a callback field for the specific action.
+
+First create a new type struct, we will call it `EnterEntry`. Then embed `widget.Entry` and insert a callback field called `OnEnter`.
+
+```go
+type EnterEntry struct {
+	widget.Entry
+
+	OnEnter func()
+}
+```
+
+As mentioned in [Extending existing widgets](https://fyne.io/develop/extending-widgets.html), we follow good practice  and create a constructor function and extend the `BaseWidget`.
+
+```go
+func newEnterEntry() *EnterEntry {
+	entry := &EnterEntry{}
+	entry.ExtendBaseWidget(entry)
+	return entry
+}
+```
+
+Then override the `KeyDown()` function that's part of `widget.Entry` by adding the following block of code into our source file. Inside this function, we will use a conditional to check if the `key.Name` value matches the `Return` key but using the `fyne.KeyReturn` variable (we could just as well have checked that key.Name == `"Return"`). This can easily be extended to check for other keys in the future if necessary.
+
+```go
+func (entry *EnterEntry) KeyDown(key *fyne.KeyEvent) {
+	switch key.Name {
+	case fyne.KeyReturn:
+		entry.OnEnter()
+	}
+}
+```
+
+Now things get interesting, we simply call our callback `OnEnter`. It doesn't do anything yet, but that's because we need to tell it  what to do using the `OnEnter` callback.
+
+Once we constuct our `EnterEntry`, we assign an anonymous function to `OnEnter`. It's important to do it this way because anonymous functions (or closures) can capture their surrounding environment of variables. This offers more flexibility to interact with more than one widget, such as passing text to a label and insert in a VBox to simulate chat history.
+
+```go
+messages := widget.NewVBox()
+
+entry := newEnterEntry()
```

Comment:
I like that this is a full illustration, but to keep the tutorial on topic perhaps we could just log “Enter pressed - content was ...” then clear it, or similar?

Reply:
The latest commit simplifies stuff a lot, I think this might be fixed.